File: plugins/line/proj.ios/ShareLine.m
/****************************************************************************
 Copyright (c) 2013 cocos2d-x.org
 
 http://www.cocos2d-x.org
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/

#import "ShareLine.h"
#import "ShareWrapper.h"
#import <AssetsLibrary/AssetsLibrary.h>

#define OUTPUT_LOG(...)     if (self.debug) NSLog(__VA_ARGS__);

@implementation ShareLine

@synthesize mShareInfo;
@synthesize debug = __debug;

- (void) configDeveloperInfo : (NSMutableDictionary*) cpInfo
{
}

- (void) share: (NSMutableDictionary*) shareInfo
{
    self.mShareInfo = shareInfo;
    //NSString* text = [NSString stringWithFormat:@"%@", [mShareInfo objectForKey:@"SharedText"]];
    NSString* imgPath = [mShareInfo objectForKey:@"SharedImagePath"];
    if (imgPath) {
        UIPasteboard* pb = [UIPasteboard generalPasteboard];
        [pb setData:UIImagePNGRepresentation([UIImage imageWithContentsOfFile:imgPath]) forPasteboardType:@"public.png"];
        NSString *urlString = [NSString stringWithFormat:@"line://msg/image/%@", pb.name];
        NSURL *url = [NSURL URLWithString:urlString];
        [[UIApplication sharedApplication] openURL:url];
    }
}

- (void) setDebugMode: (BOOL) debug
{
    self.debug = debug;
}

- (NSString*) getSDKVersion
{
    return @"0";
}

- (NSString*) getPluginVersion
{
    return @"0.0.1";
}

- (void) saveImageToGallery: (NSMutableDictionary*) params
{
    NSString* imgPath = [params objectForKey:@"imagePath"];
    
    //iOS6以上ではフォトアルバムに画像を保存する時アクセス許可が必要なのでチェック処理
    BOOL isAuthorization = YES;
    NSArray  *aOsVersions = [[[UIDevice currentDevice]systemVersion] componentsSeparatedByString:@"."];
    NSInteger iOsVersionMajor  = [[aOsVersions objectAtIndex:0] intValue];
    if (iOsVersionMajor >= 6)
    {
        //このアプリの写真への認証状態を取得する
        isAuthorization = NO;
        switch ([ALAssetsLibrary authorizationStatus])
        {
            case ALAuthorizationStatusNotDetermined:
            case ALAuthorizationStatusAuthorized:
            {
                //写真へのアクセスを許可するか選択されていない
                //写真へのアクセスが許可されている
                isAuthorization = YES;
                break;
            }
            case ALAuthorizationStatusRestricted:
            {
                //設定 > 一般 > 機能制限で利用が制限されている
                isAuthorization = NO;
                UIAlertView *alertView = [[UIAlertView alloc]
                                          initWithTitle:@"エラー"
                                          message:@"写真へのアクセスが許可されていません。\n設定 > 一般 > 機能制限で許可してください。"
                                          delegate:nil
                                          cancelButtonTitle:@"OK"
                                          otherButtonTitles:nil];
                [alertView show];
                break;
            }
            case ALAuthorizationStatusDenied:
            {
                //設定 > プライバシー > 写真で利用が制限されている
                isAuthorization = NO;
                UIAlertView *alertView = [[UIAlertView alloc]
                                          initWithTitle:@"エラー"
                                          message:@"写真へのアクセスが許可されていません。\n設定 > プライバシー > 写真で許可してください。"
                                          delegate:nil
                                          cancelButtonTitle:@"OK"
                                          otherButtonTitles:nil];
                [alertView show];
                break;
            }
            default:
                break;
        }
    }
    
    if ( isAuthorization && imgPath ) {
        UIPasteboard* pb = [UIPasteboard generalPasteboard];
        [pb setData:UIImagePNGRepresentation([UIImage imageWithContentsOfFile:imgPath]) forPasteboardType:@"shobon_screenshot.png"];
        
        ALAssetsLibrary *assetsLibrary = [ALAssetsLibrary new];
        [assetsLibrary
         writeImageDataToSavedPhotosAlbum:UIImagePNGRepresentation([UIImage imageWithContentsOfFile:imgPath])
         metadata:nil
         completionBlock:^(NSURL *assetURL, NSError *error) {
            if (error) {
                UIAlertView *alertView = [[UIAlertView alloc]
                                          initWithTitle:@"エラー"
                                          message:@"写真への保存に失敗しました。"
                                          delegate:nil
                                          cancelButtonTitle:@"OK"
                                          otherButtonTitles:nil];
                [alertView show];
            } else {
                UIAlertView *alertView = [[UIAlertView alloc]
                                          initWithTitle:@""
                                          message:@"保存しました。"
                                          delegate:nil
                                          cancelButtonTitle:@"OK"
                                          otherButtonTitles:nil];
                [alertView show];
            }
        }];
    }
}

@end

File: plugins/line/proj.android/src/org/cocos2dx/plugin/ShareLine.java
/****************************************************************************
Copyright (c) 2012-2013 cocos2d-x.org

http://www.cocos2d-x.org

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
****************************************************************************/
package org.cocos2dx.plugin;

import java.util.Hashtable;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.util.Log;
import android.provider.MediaStore;
import android.app.AlertDialog;
import android.content.DialogInterface;
import java.io.File;

import org.json.JSONObject;

public class ShareLine implements InterfaceShare {

    private static final String LOG_TAG = "ShareLine";
    private static Activity mContext = null;
    private static InterfaceShare mShareAdapter = null;
    protected static boolean bDebug = false;

    private static Hashtable<String, String> mShareInfo = null;

    public static String KEY_TEXT="SharedText";
    public static String KEY_IMAGE_PATH = "SharedImagePath";

    protected static void LogE(String msg, Exception e) {
        Log.e(LOG_TAG, msg, e);
        e.printStackTrace();
    }

    protected static void LogD(String msg) {
        if (bDebug) {
            Log.d(LOG_TAG, msg);
        }
    }

    public ShareLine(Context context) {
      mContext = (Activity) context;
      mShareAdapter = this;
    }

    @Override
    public void configDeveloperInfo(Hashtable<String, String> cpInfo) {
        LogD("initDeveloperInfo invoked " + cpInfo.toString());
    }

    @Override
    public void share(Hashtable<String, String> info) {
        LogD("share invoked " + info.toString());
        mShareInfo =  info;
        PluginWrapper.runOnMainThread(new Runnable() {
            @Override
            public void run() {
                String text = mShareInfo.get(KEY_TEXT);
                String imagePath = mShareInfo.get(KEY_IMAGE_PATH);
                
                File f = new File(imagePath);
                if(f.exists()) {
                    //他のアプリからこの画像を読めるようにする
                    f.setReadable(true, false);
                    
                    Intent intent = new Intent(Intent.ACTION_VIEW);
                    intent.setData(Uri.parse("line://msg/image/" + imagePath));
                    mContext.startActivity(intent);
                }
            }
        });
    }

    @Override
    public void setDebugMode(boolean debug) {
        bDebug = debug;
    }

    @Override
    public String getSDKVersion() {
        return "Unknown version";
    }

    @Override
    public String getPluginVersion() {
        return "0.0.1";
    }
    
    public void saveImageToGallery(JSONObject params) {
        try {
            final String imgPath = params.getString("imagePath");
            
            PluginWrapper.runOnMainThread(new Runnable() {
                @Override
                public void run() {
                    
                    String uriStr;
                    try {
                        uriStr = MediaStore.Images.Media.insertImage(mContext.getContentResolver(), imgPath, "", "");
                        if(uriStr!=null) {
                            //保存成功
                            AlertDialog.Builder dialog = new AlertDialog.Builder(mContext);
                            dialog.setTitle("");
                            dialog.setMessage("保存しました。");
                            dialog.setPositiveButton("OK",new DialogInterface.OnClickListener() {
                                public void onClick(DialogInterface dialog,int whichButton) {
                                }
                            });
                            dialog.create();
                            dialog.show();
                            
                        } else {
                            //保存失敗
                            AlertDialog.Builder dialog = new AlertDialog.Builder(mContext);
                            dialog.setTitle("エラー");
                            dialog.setMessage("保存に失敗しました。");
                            dialog.setPositiveButton("OK",new DialogInterface.OnClickListener() {
                                public void onClick(DialogInterface dialog,int whichButton) {
                                }
                            });
                            dialog.create();
                            dialog.show();
                        }
                        
                    }  catch (Exception e) {
                        //エラー
                        AlertDialog.Builder dialog = new AlertDialog.Builder(mContext);
                        dialog.setTitle("エラー");
                        dialog.setMessage("保存に失敗しました。");
                        dialog.setPositiveButton("OK",new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog,int whichButton) {
                            }
                        });
                        dialog.create();
                        dialog.show();
                        LogE("Media.insertImage error", e);
                    }
                }
            });
        } catch (org.json.JSONException e) {
            LogE("invalid param", e);
        }
    }

}
